### PR TITLE
PRIME-2180 Out of Province College License

### PR DESCRIPTION
**Pull request by Paahn**
- go to admin and enter Paper Enrolment
- fill out forms and when you reach ` /regulatory` the new toggle is available

### CODE REVIEW

**Comment by neophyte57** [comment]
Each time I toggle the control to "on", it adds a new set of out-of-province college fields, probably not desired behavior 
image

**Comment by neophyte57** [comment]
I think we may need some front-end validation 

image

**Comment by neophyte57** [comment]
When I enter information like this 
image

On "Save and Continue", I see the "Site Information" screen.  Although "None" typical license was selected, I think having Out-of-Province college information should cause the "Site Information" screen to be bypassed?  Also are all Out-of-Province College Codes 5 digits?      

When I go "Back", the toggle returns to the off state, although the previously entered Out-of-Province college information is visible when the toggle is back to "on" state.

image

File: prime-angular-frontend/src/app/shared/components/forms/unlisted-college-licence-form/unlisted-college-licence-form.component.html [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      </div>
+    </div>
+    <div class="row">
+      <div class="col">
+
+        <mat-form-field class="w-100">
+          <input matInput
+                 placeholder="College Code"
+                 formControlName="unlistedCollegeCode"
+                 maxlength="30">
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/shared/components/forms/unlisted-college-licence-form/unlisted-college-licence-form.component.html:26]: A few things, (1) got this error when provided "E5" as input 

image

(2) The mock-up names this field differently:  https://xd.adobe.com/view/2c575e9f-46f3-4c99-a0e6-89be106407a6-7904/screen/8790f734-8454-4eeb-9692-70677e95f65e/ 

(3) Please check the mock-up for length limitations for the fields

File: prime-angular-frontend/src/app/modules/paper-enrolment/pages/regulatory-page/regulatory-overview.component.ts [hunk 2]
```diff
@@ -73,6 +73,47 @@ import { RegulatoryForm } from './regulatory-form.model';
...
+                mat-icon-button
+                matTooltip="Edit College Licences"
+                (click)="onRoute(PaperEnrolmentRoutes.REGULATORY)">
+          <mat-icon>edit</mat-icon>
+        </button>
+      </app-page-subheader>
+
+
+      <ng-container *ngIf="!regulatory?.unlistedCertifications?.length; else unlistedCerts">
+        No unlisted certifications entered.
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/paper-enrolment/pages/regulatory-page/regulatory-overview.component.ts:92]: Can you please check with someone from the Business whether this text makes sense to them?

File: prime-angular-frontend/src/app/shared/components/enrollee/enrollee-review/enrollee-review.component.html [hunk 3]
```diff
@@ -246,6 +246,46 @@
...
+    <button *ngIf="showEditRedirect"
+            mat-icon-button
+            matTooltip="Edit College Licences"
+            (click)="onRoute(EnrolmentRoutes.REGULATORY)">
+      <mat-icon>edit</mat-icon>
+    </button>
+  </app-page-subheader>
+
+  <ng-container *ngIf="!enrolment?.unlistedCertifications?.length; else unlistedCerts">
+    No unlisted certifications entered.
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/shared/components/enrollee/enrollee-review/enrollee-review.component.html:263]: Can you please check with someone from the Business whether this text makes sense to them?

File: prime-dotnet-webapi/Models/Certifications/UnlistedCertification.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    public class UnlistedCertification : BaseAuditable, IEnrolleeNavigationProperty
+    {
+        [Key]
+        public int Id { get; set; }
+        [JsonIgnore]
+        public int EnrolleeId { get; set; }
+        public string UnlistedCollegeName { get; set; }
+        public string UnlistedCollegeCode { get; set; }
+        public DateTimeOffset? UnlistedRenewalDate { get; set; }
+        public Enrollee Enrollee { get; set; }
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/Certifications/UnlistedCertification.cs:18]: Do you need "Unlisted" as part of the field names, considering the class is named `UnlistedCertification`?

**Comment by bergomi02** [comment]
In Test in PR, I have some issue when submitting a paper enrolment, it said "enrolment can not be submitted" or something like that. There is the console output. 
image
However, I clicked the submit button again and refresh the page then it shows the submission is successful with green banner. Not sure if it is a one-time thing.

Other thing I noticed is that, the Other College Licence is not showing under Enrolment -> View Enrolment.
image
image
If I just go to Overview, the Other College Licence entry is showing up correctly.
image

**Comment by neophyte57** [comment]
image

Can the "College Licence" field be indented, like "Next Renewal Date" as I think they conceptually belong under "College Name"?

**Comment by neophyte57** [comment]
> In Test in PR, I have some issue when submitting a paper enrolment, it said "enrolment can not be submitted" or something like that. There is the console output. image However, I clicked the submit button again and refresh the page then it shows the submission is successful with green banner. Not sure if it is a one-time thing.
> 
> Other thing I noticed is that, the Other College Licence is not showing under Enrolment -> View Enrolment. image image If I just go to Overview, the Other College Licence entry is showing up correctly. image

This still seems to be a problem, that is, I can see the Unlisted Certification in "Overview" but not through "Enrolments".  On the latter screen, it just says "No unlisted certifications entered."  For example, see this enrollment in the PR environment:

image

**Comment by Paahn** [comment]
> > In Test in PR, I have some issue when submitting a paper enrolment, it said "enrolment can not be submitted" or something like that. There is the console output. image However, I clicked the submit button again and refresh the page then it shows the submission is successful with green banner. Not sure if it is a one-time thing.
> > Other thing I noticed is that, the Other College Licence is not showing under Enrolment -> View Enrolment. image image If I just go to Overview, the Other College Licence entry is showing up correctly. image
> 
> This still seems to be a problem, that is, I can see the Unlisted Certification in "Overview" but not through "Enrolments". On the latter screen, it just says "No unlisted certifications entered." For example, see this enrollment in the PR environment:
> 
> image

@neophyte57 I see what happens, other college licence information doesnt show up for paper enrolments submitted before [this commit](https://github.com/bcgov/moh-prime/pull/2154/commits/51a3f87db4ba940267e3413469717c953fbf53bc) as the enrolments page displays submissions, and before the commit/fix enrolment submission is created without the Other college information data. We can discuss a solution but given that Paper enrolments cannot be edited at this time not sure its worth it?

However I keep getting the submission error that @bergomi02 noticed as well on the PR environment but not on local, not sure why.

File: prime-dotnet-webapi/Models/Certifications/UnlistedCertification.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+{
+    [Table("UnlistedCertification")]
+    public class UnlistedCertification : BaseAuditable, IEnrolleeNavigationProperty
+    {
+        [Key]
+        public int Id { get; set; }
+        [JsonIgnore]
+        public int EnrolleeId { get; set; }
+        public string CollegeName { get; set; }
+        public string CollegeCode { get; set; }
```
[the remark this answers is not in the document]
Comment (Paahn) [reply prime-dotnet-webapi/Models/Certifications/UnlistedCertification.cs:16]: Good observation, I just noticed the other use of `CollegeCode` in the app. Would it make more sense to name this field `LicenceCode` or `CollegeLicence` ?

**Comment by neophyte57** [comment]
> > > In Test in PR, I have some issue when submitting a paper enrolment, it said "enrolment can not be submitted" or something like that. There is the console output. image However, I clicked the submit button again and refresh the page then it shows the submission is successful with green banner. Not sure if it is a one-time thing.
> > > Other thing I noticed is that, the Other College Licence is not showing under Enrolment -> View Enrolment. image image If I just go to Overview, the Other College Licence entry is showing up correctly. image
> > 
> > 
> > This still seems to be a problem, that is, I can see the Unlisted Certification in "Overview" but not through "Enrolments". On the latter screen, it just says "No unlisted certifications entered." For example, see this enrollment in the PR environment:
> > image
> 
> @neophyte57 I see what happens, other college licence information doesnt show up for paper enrolments submitted before [this commit](https://github.com/bcgov/moh-prime/pull/2154/commits/51a3f87db4ba940267e3413469717c953fbf53bc) as the enrolments page displays submissions, and before the commit/fix enrolment submission is created without the Other college information data. We can discuss a solution but given that Paper enrolments cannot be edited at this time not sure its worth it?
> 
> However I keep getting the submission error that @bergomi02 noticed as well on the PR environment but not on local, not sure why.

I wouldn't worry about the 504 Gateway Time-out issue, I think it's due to the poor performance of the PR environments @bergomi02  

OK, don't worry about the old submissions in the PR environments missing unlisted certifications.  I see that submissions since your commit/fix display the unlisted certifications in the "Enrolments" section.

All good!  @Paahn

File: prime-dotnet-webapi/Models/Certifications/UnlistedCertification.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (neophyte57) [reply prime-dotnet-webapi/Models/Certifications/UnlistedCertification.cs:16]: Yeah, you're right, `LicenseNumber` would be better, like in the `Certification` table

**Comment by bergomi02** [comment]
Getting server error when trying to save the first step of a paper enrolment:
image
image

File: prime-angular-frontend/src/app/shared/components/enrollee/enrollee-review/enrollee-review.component.html (hunk 3 above)
Comment (Paahn) [reply prime-angular-frontend/src/app/shared/components/enrollee/enrollee-review/enrollee-review.component.html]: @neophyte57 @bergomi02 received the text today from the business and updated the code.

File: prime-angular-frontend/src/app/modules/paper-enrolment/pages/regulatory-page/regulatory-overview.component.ts (hunk 2 above)
Comment (Paahn) [reply prime-angular-frontend/src/app/modules/paper-enrolment/pages/regulatory-page/regulatory-overview.component.ts]: Text received from business and updated

**Pull request merged**